### ISSUE TITLE
research: investigate suitability of sensitivity analysis for param sweep

### ISSUE THREAD

**Issue by mikkokotila** [Vaquum/Limen#213]
# Sensitivity Analysis for Parameter Space Pruning

## Investigation Goal

**Question:** What sensitivity analysis methods are suitable for Limen's parameter sweeping, and how can they inform intelligent pruning strategies to reduce search space without sacrificing performance?

**Context:** Limen experiments sweep large parameter spaces (often millions of combinations) to find optimal model configurations. Many parameters may have negligible impact on performance, wasting compute resources. Sensitivity analysis can identify which parameters matter most, enabling pruning strategies to fix or narrow low-impact parameters and focus exploration on high-impact dimensions.

---
## Overview of Sensitivity Analysis for Parameter Sweeping

Sensitivity analysis quantifies how much each input parameter influences output metrics. In Limen's context:

- **Inputs (X):** Model parameters from experiment log (e.g., `shift`, `C`, `max_iter`, `tol`, `solver`)
- **Output (Y):** Performance metrics (e.g., `auc`, `accuracy`, `trade_win_rate_pct`)
- **Goal:** Identify parameters with low sensitivity → candidates for pruning

**Key Challenge:** Limen uses different search strategies (Random, Grid, TPE, Adaptive) with varying sampling characteristics. Sensitivity methods have different assumptions about data distribution, making some methods more suitable than others depending on the search strategy used.

**Three Methods Evaluated:**

1. **Correlation-based Analysis**  (Already implemented in Limen)
2. **Random Forest Feature Importance**
3. **Sobol Variance Decomposition**

---

### Description

Trains a Random Forest model using parameters as features and target metric as output, then computes permutation importance to quantify each parameter's contribution to prediction accuracy.

**Implementation:** New - requires scikit-learn (already available in Limen environment)

**How it works:**
```python
from sklearn.ensemble import RandomForestRegressor
from sklearn.inspection import permutation_importance

# 1. Prepare data from experiment log
X = log_df[param_cols].values  # Parameters
y = log_df['auc'].values       # Target metric

# 2. Train Random Forest
rf = RandomForestRegressor(n_estimators=100, random_state=42)
rf.fit(X, y)

# 3. Compute permutation importance
perm_importance = permutation_importance(rf, X, y, n_repeats=10)

# Returns: importance scores + std for each parameter
```

**Key Features:**
- Model-based sensitivity (no distribution assumptions)
- Permutation importance: shuffle parameter values, measure prediction degradation
- Native support for continuous and categorical parameters
- Captures non-linear relationships and interactions implicitly
- Provides uncertainty estimates via repeated permutations

### Applicability to Limen

**Strengths:**

- Works with ANY search strategy - Random, Grid, TPE, Adaptive - no uniformity assumption
- Handles non-linearity - Captures complex parameter effects ML hyperparameters exhibit
- Fast computation
- Robust to noise - Ensemble averaging reduces sensitivity to outliers
- Mixed parameter types - Handles continuous and categorical without preprocessing
- Implicit interaction capture - Tree splits naturally model parameter interactions

**Limitations:**

- Black-box method - Less theoretically grounded than variance decomposition
- Biased toward high-cardinality - Parameters with many values may appear more important
- No explicit interaction decomposition - Can't separate main effects from interactions
- Requires careful tuning - RF hyperparameters affect importance estimates

### Integration with Pruning Strategy

**Data preparation:**
1. Extract parameter columns from experiment log, excluding metrics (auc, accuracy, etc.) and metadata (id, execution_time)
2. Encode categorical parameters using LabelEncoder to convert strings/categories to numeric codes
3. Keep numeric parameters as-is without transformation
4. Maintain parameter names for interpretation
5. Check minimum sample size requirement before proceeding

**Model training:**
1. Train RandomForestRegressor with experiment log parameters as features (X) and target metric as output (y)
2. Use reasonable defaults: 100 trees, parallel processing (n_jobs=-1), fixed random state for reproducibility

**Importance computation:**
1. Use permutation importance rather than tree-based importance for more reliable estimates
2. Shuffle each parameter's values independently, measure prediction degradation
3. Repeat permutations `n` times to estimate uncertainty (standard deviation)
4. Returns mean importance score and standard deviation for each parameter

**Parameter classification:**
1. Identify low-importance parameters where mean importance falls below threshold (e.g., < 0.001)
2. Consider standard deviation - high variance suggests unreliable estimate, may want to skip pruning
3. Rank parameters by importance for decision-making

**Intervention generation:**
1. For each low-importance parameter, determine a fixed value to use
2. Options: median from all data, mode from top ` k`% performers, or most common value overall
3. Generate `keep_is` intervention with chosen value
4. Include importance score and standard deviation in intervention metadata for audit trail

**Best suited for:** Search strategy that does not guarantee sampling with `uniform distribution` such as TPE, Adaptive TPE.

---

### Description

Variance-based global sensitivity analysis that decomposes output variance into contributions from each parameter (first-order) and their interactions (total-order), using Sobol indices.

**Implementation:** New - requires scipy.stats.sobol_indices (available in scipy ≥1.10)

**How it works:**
```python
from scipy.stats import sobol_indices

# 1. Select target metric (output Y)
y = log_df['auc'].values  # The metric we want to understand

# 2. Prepare input parameters (X) - exclude target and other metrics
exclude_cols = ['id', 'execution_time', 'auc', 'accuracy', 'precision', 'recall', 'fpr']
param_cols = [c for c in log_df.columns if c not in exclude_cols]
X = log_df[param_cols].values  # Only numeric input parameters

# 3. Normalize parameters to [0, 1] (Sobol requirement)
X_normalized = (X - X.min(axis=0)) / (X.max(axis=0) - X.min(axis=0))

# 4. Define parameter distributions
dists = {param_cols[i]: X_normalized[:, i] for i in range(len(param_cols))}

# 5. Compute Sobol indices
# "How much variance in auc (Y) is explained by each parameter (X)?"
result = sobol_indices(
    func=None,  # Use existing samples
    n=len(y),
    dists=dists,
    method='saltelli'
)

# Returns:
# - first_order: S_i (variance from parameter i alone)
# - total_order: S_Ti (variance from parameter i + interactions)
# - interaction_effect: S_Ti - S_i (variance from interactions only)
```

**Key Features:**
- Rigorous variance decomposition with theoretical foundation
- Separates main effects from interaction effects explicitly
- First-order: parameter's isolated contribution
- Total-order: parameter + all interactions involving it
- Interaction strength: total - first order

### Applicability to Limen

**Strengths:**

- Theoretically rigorous - Grounded in variance decomposition theory
- Interaction decomposition - Explicitly quantifies interaction effects
- Global sensitivity - Considers entire parameter space, not local gradients
- Interpretable - Indices sum to 1.0 (percentage of variance explained)
- Publication-ready - Standard method in sensitivity analysis literature
- Identifies important interactions - Can detect when parameters only matter in combination

**Limitations:**

- Requires uniform sampling - Assumes parameters sampled uniformly or quasi-uniformly
- Breaks with adaptive strategies - TPE/Adaptive bias toward good regions violates assumptions
- Computationally expensive - O(n·p²), approximately 10x slower than Random Forest
- Assumes independence - Correlated parameters complicate interpretation
- Categorical handling - Requires one-hot encoding, inflates dimensionality
- Large sample requirement - Needs 500-1000+ samples for reliable estimates

**Suitability for Parameter Sweeping:**

**GOOD for:** Experiments using Random, Grid, or Sobol search strategies where uniform sampling is guaranteed. Provides richer insights than RF when assumptions are met.

**NOT SUITABLE for:** Experiments using TPE or Adaptive strategies. The biased sampling toward high-performing regions violates the uniform sampling assumption, making Sobol indices potentially misleading (may underestimate importance of parameters that perform poorly in the biased sample).

### Integration with Pruning Strategy

**Prerequisite validation:**
1. Check minimum sample size requirement (typically 500+ rounds) before attempting Sobol analysis
2. Validate sampling uniformity using Kolmogorov-Smirnov test against uniform distribution
3. Require at least 70% of parameters to pass uniformity test (p-value > 0.05)
4. If uniformity check fails, reject Sobol method and recommend Random Forest instead
5. Only proceed if using Random, Grid, or Sobol search strategies

**Data preparation:**
1. Select target metric as output variable Y (e.g., `auc`) - this is what we're analyzing sensitivity FOR
2. Extract only numeric input parameters as X, excluding:
   - The target metric itself (`auc`)
   - Other output metrics (`accuracy`, `precision`, `recall`, `fpr`)
   - Metadata columns (`id`, `execution_time`)
3. Normalize all input parameters X to [0, 1] range using min-max normalization (required by Sobol)
4. Categorical parameters require one-hot encoding if included, which increases dimensionality

**Key distinction:** X contains only numeric input parameters (shift, C, max_iter), Y is the output metric (auc). Sobol answers: "How much variance in Y is caused by each parameter in X?"

**Parameter classification:**
1. Low-sensitivity parameters: total-order index below threshold (e.g., < 0.05), candidates for pruning
2. High-sensitivity parameters: total-order index above threshold, keep full range
3. Interaction-heavy parameters: large difference between total and first-order (e.g., > 0.1), important through interactions, flag but don't prune
---

### The Multi-Metric Problem

In practice, experiments often optimize multiple objectives simultaneously (e.g., `auc`, `accuracy`, `trade_win_rate_pct`, `max_drawdown_pct`). This raises the question: **How do we perform sensitivity analysis when we care about multiple output metrics?**

The methods described above (Correlation, Random Forest, Sobol) assume a **single scalar output Y**. For multiple metrics, we need extended approaches.

### Approach 2: Multi-Output Random Forest

**Concept:** Train a single Random Forest model that predicts multiple outputs simultaneously.

```python
from sklearn.ensemble import RandomForestRegressor

# Multiple metrics as vector output
y_multi = log_df[['auc', 'accuracy', 'trade_win_rate_pct']].values  # Shape: (n_samples, 3)
X = log_df[param_cols].values

# Train multi-output RF (sklearn handles this natively)
rf = RandomForestRegressor(n_estimators=100)
rf.fit(X, y_multi)

# Get aggregated importance (averaged across all outputs)
importance_avg = rf.feature_importances_

# Can also compute per-output importance if needed
```

**Advantages:**
- Single model captures joint relationships across all metrics
- Efficient: one training instead of N separate models
- Identifies parameters important for multiple objectives
- Native support in scikit-learn (no extra dependencies)
- Can extract both averaged AND per-output importance

**Disadvantages:**
- Averaged importance may hide metric-specific effects
- Less interpretable than separate analyses
- Requires more samples for reliable estimates (more outputs = more complexity)
- Only available for Random Forest (not Correlation or Sobol)

**Pruning decision:** Use averaged importance to identify parameters consistently unimportant across all metrics.
---
### Approach 3: Composite Metric

**Concept:** Create a weighted combination of metrics and analyze sensitivity to the composite.

```python
# Define multi-objective preferences
weights = {'auc': 0.5, 'accuracy': 0.2, 'trade_win_rate_pct': 0.3}

# Create composite scalar metric
y_composite = (
    0.5 * normalize(log_df['auc']) +
    0.2 * normalize(log_df['accuracy']) +
    0.3 * normalize(log_df['trade_win_rate_pct'])
)

# Single sensitivity analysis
importance = compute_sensitivity(X, y_composite)
```

**Advantages:**
- Single analysis with clear optimization target
- All three methods (Correlation, RF, Sobol) work
- Captures user's multi-objective preferences explicitly
- Simple to implement

**Disadvantages:**
- Requires defining weights (subjective)
- Loses information about metric-specific sensitivities
- Assumes linear combination is meaningful
- Different weight choices yield different results

**Pruning decision:** Parameters with low composite sensitivity can be pruned.

**Supported by:**
- Random Forest: Works with any scalar target
- Sobol Indices:  Works with any scalar target

### Approach 4: Generalized Sobol Indices (Research-Level)

**Concept:** True multivariate Sobol indices that consider covariance structure between outputs.

**Theory:** Extends variance decomposition to vector-valued functions, decomposing total variance-covariance matrix.

**Implementation:**
- **UQLab (MATLAB):** Full support for generalized Sobol indices for vector outputs
  - Computes generalized first-order and total-order indices
  - Considers correlations between output metrics
  - Theoretically rigorous, publication-ready
  - **Problem:** Requires MATLAB license and MATLAB Engine for Python

- **scipy:**  Not supported (scalar output only)
- **SALib:**  Not natively supported (must loop through outputs)

**Advantages:**
- Theoretically rigorous variance decomposition
- Captures correlations between outputs
- Standard method in multi-objective optimization research
- Publication-ready results

**Disadvantages:**
- Not available in Python ecosystem (UQLab is MATLAB-only)
- MATLAB licensing cost ($2k-5k) and deployment complexity
- Still requires uniform sampling (same restriction as scalar Sobol)
- Integration overhead (Python ↔ MATLAB)
- Overkill for most practical applications
---
### Multi-Output Support Summary

| Method | Single Output | Multi-Output (Separate) | Multi-Output (Native) | Multi-Output (Composite) |
|--------|--------------|------------------------|----------------------|-------------------------|
| **Correlation** | ✅ Default | ✅ Loop through metrics | ❌ N/A | ✅ Works |
| **Random Forest** | ✅ Default | ✅ Loop through metrics | ✅ Native sklearn support | ✅ Works |
| **Sobol (scipy)** | ✅ Default | ✅ Loop through metrics | ❌ Scalar only | ✅ Works |
| **Sobol (UQLab)** | ✅ Default | ✅ Loop through metrics | ✅ Generalized Sobol | ✅ Works |

---
### Implementation Notes

**For Random Forest:** Multi-output is straightforward with sklearn - use as primary method.

**For Sobol:** Limited to separate analysis per metric. Accept this limitation since Sobol already has uniform sampling restriction. Run for primary metric only, or loop through metrics if needed.

**For Correlation:** Already fast, running separately for each metric adds minimal overhead.

**Safety principle:** When in doubt, don't prune. Only remove parameters that are consistently unimportant across ALL objectives.

---
## Conclusion: Suitability Assessment

**Is sensitivity analysis suitable for Limen's parameter sweeping?**

**YES**, with method selection based on search strategy:

**Random Forest** is highly suitable for all Limen experiments:
- Works regardless of search strategy
- Fast enough for periodic feedback
- Proven in AutoML systems
- Already have dependencies

**Correlation Analysis** is suitable for early screening:
- Already implemented
- Very fast
- Good for obvious cases
- Complements other methods

**Sobol Indices** are suitable for subset of experiments:
- Only works with uniform sampling (Random/Grid/Sobol strategies)
- Provides richer insights when applicable (explicit interaction decomposition)
- Optional enhancement, not required

**Recommended Implementation Roadmap:**

**Phase 1** (Immediate): Use existing correlation analysis for early screening

**Phase 2** (Next sprint): Implement Random Forest sensitivity pruning as primary method

**Phase 3** (Future enhancement): Add Sobol indices for experiments using uniform sampling strategies
